### PR TITLE
Update refos to master

### PR DESCRIPTION
**Pull request by ratmice**
This updates refos to master on adding onto hypernewbie's pull request
it also updates the libraries to use the seL4_libs/util_libs/seL4_tools repositories

[edit]... I'm not currently set up to test the ARM configs, so they may need changes.

### CODE REVIEW

File: impl/apps/process_server/src/state.c [hunk 1]
```diff
@@ -63,28 +64,30 @@ initialise_welcome_message(seL4_BootInfo *info)
...
     dprintf("Untyped Slot       Paddr      Bits\n");
     int numUntyped = info->untyped.end - info->untyped.start;
+    int numDevices = 0;
     for (int i = 0; i < numUntyped; i++) {
-        dprintf("%3x     0x%08x 0x%08x %d\n", i, info->untyped.start+i, info->untypedPaddrList[i],
-                info->untypedSizeBitsList[i]);
+        if (!info->untypedList[i].isDevice) {
+          dprintf("%3x     0x%08x 0x%08x %d\n", i, info->untyped.start+i, info->untypedList[i].paddr,
+                  info->untypedList[i].sizeBits);
+        } else numDevices++;
```
Comment (hypernewbie) [inline impl/apps/process_server/src/state.c:78]: Prefer for this to be on new line { else block.
Not a big thing.

File: impl/apps/process_server/src/system/process/thread.c [hunk 2]
```diff
@@ -57,16 +64,34 @@ thread_config(struct proc_tcb *thread, uint8_t priority, vaddr_t entryPoint,
...
 thread_start(struct proc_tcb *thread, void *arg0, void *arg1)
 {
+    int error;
     assert(thread);
     assert(thread->entryPoint);
 
     /* Start the thread using seL4utils library. */
-    int error = sel4utils_internal_start_thread (
+#ifdef CONFIG_KERNEL_MASTER
+    seL4_UserContext userContext = {0};
```
Comment (hypernewbie) [inline impl/apps/process_server/src/system/process/thread.c:73]: Comment here saying why master kernel is different would be appreciated!

File: impl/libs/librefos/src/refos-rpc/rpc.c [hunk 3]
```diff
@@ -143,14 +143,14 @@ rpc_init(const char* name_str, int32_t label)
...
-    seL4_SetMR(0, label);
-
     if (!_rpc_recv_cslot) {
         rpc_setup_recv(REFOS_THREAD_CAP_RECV);
     } else if (seL4_MessageInfo_get_extraCaps(_rpc_minfo) > 0) {
         // Flush recieving path of previous recieved caps.
         seL4_CNode_Delete(REFOS_CSPACE, _rpc_recv_cslot, REFOS_CDEPTH);
     }
+
+    seL4_SetMR(0, label);
```
Comment (hypernewbie) [inline impl/libs/librefos/src/refos-rpc/rpc.c:153]: Why was this change made?

**Review by hypernewbie — APPROVED** [review]
LGTM

File: impl/libs/librefos/src/refos-rpc/rpc.c (hunk 3 above)
Comment (ratmice) [reply impl/libs/librefos/src/refos-rpc/rpc.c:153]: when doing a system call, the second system call was always being corrupted
I surmised that in the else block, the seL4_CNode_Delete was somehow deleting the label which was put in the seL4_SetMR call.

thanks for calling this out as perhaps it needs a different fix.

Comment (hypernewbie) [reply impl/libs/librefos/src/refos-rpc/rpc.c:153]: ok. this is acceptable as a WAR but please try to find root cause

Comment (AdrianDanis) [reply impl/libs/librefos/src/refos-rpc/rpc.c:153]: `seL4_CNode_Delete` previously would have used an optimized syscall stub that would send a message without bother to place the message into the IPC buffer, but would instead place it straight into register. For some silly reasons using these optimized stubs causes parts of kernel error messages to get thrown away. So currently invocations use the very slightly less efficient stubs that load and store the from the IPC buffer, thus overwriting the label.

tl;dr: label should always have been set after, it working before was by chance not design

Comment (hypernewbie) [reply impl/libs/librefos/src/refos-rpc/rpc.c:153]: thanks for clearing up

**Comment by AdrianDanis** [comment]
Our internal work on refOS has now been pushed out here if you want to see what changes of yours overlap etc.

**Comment by ratmice** [comment]
Thanks for the heads up, I've created a new pull request with the fallout, closing
